lightning/src/ln/payment_tests.rs

```diff
@@ -2748,3 +2748,63 @@ fn test_threaded_payment_retries() {
 		}
 	}
 }
+
+fn do_no_missing_sent_on_midpoint_reload(persist_manager_with_payment: bool) {
+	// Test that if we reload in the middle of an HTLC claim commitment signed dance we'll still
+	// receive the PaymentSent event even if the ChannelManager had no idea about the payment when
+	// it was last persisted.
+	let chanmon_cfgs = create_chanmon_cfgs(2);
+	let node_cfgs = create_node_cfgs(2, &chanmon_cfgs);
+	let node_chanmgrs = create_node_chanmgrs(2, &node_cfgs, &[None, None]);
+	let (persister_a, persister_b);
+	let (chain_monitor_a, chain_monitor_b);
+	let (nodes_0_deserialized, nodes_0_deserialized_b);
+	let mut nodes = create_network(2, &node_cfgs, &node_chanmgrs);
+
+	let chan_id = create_announced_chan_between_nodes(&nodes, 0, 1).2;
+
+	let mut nodes_0_serialized = Vec::new();
+	if !persist_manager_with_payment {
+		nodes_0_serialized = nodes[0].node.encode();
+	}
+
+	let (our_payment_preimage, our_payment_hash, _) = route_payment(&nodes[0], &[&nodes[1]], 1_000_000);
+
+	if persist_manager_with_payment {
+		nodes_0_serialized = nodes[0].node.encode();
+	}
+
+	nodes[1].node.claim_funds(our_payment_preimage);
+	check_added_monitors!(nodes[1], 1);
+	expect_payment_claimed!(nodes[1], our_payment_hash, 1_000_000);
+
+	let updates = get_htlc_update_msgs!(nodes[1], nodes[0].node.get_our_node_id());
+	nodes[0].node.handle_update_fulfill_htlc(&nodes[1].node.get_our_node_id(), &updates.update_fulfill_htlcs[0]);
+	nodes[0].node.handle_commitment_signed(&nodes[1].node.get_our_node_id(), &updates.commitment_signed);
+	check_added_monitors!(nodes[0], 1);
+
+	// The ChannelMonitor should always be the latest version, as we're required to persist it
+	// during the commitment signed handling.
+	let chan_0_monitor_serialized = get_monitor!(nodes[0], chan_id).encode();
+	reload_node!(nodes[0], test_default_channel_config(), &nodes_0_serialized, &[&chan_0_monitor_serialized], persister_a, chain_monitor_a, nodes_0_deserialized);
+
+	let events = nodes[0].node.get_and_clear_pending_events();
+	assert_eq!(events.len(), 2);
+	if let Event::ChannelClosed { reason: ClosureReason::OutdatedChannelManager, .. } = events[0] {} else { panic!(); }
+	if let Event::PaymentSent { payment_preimage, .. } = events[1] { assert_eq!(payment_preimage, our_payment_preimage); } else { panic!(); }
+	// Note that we don't get a PaymentPathSuccessful here as we leave the HTLC pending to avoid
+	// the double-claim that would otherwise appear at the end of this test.
+
+	for _ in 0..(IDEMPOTENCY_TIMEOUT_TICKS * 2) { nodes[0].node.timer_tick_occurred(); }
```

Comment:
Could assert that `pending_outbound_payments` is empty. I'm not totally sure why we're doing this, though.